compiler/noirc_frontend/src/hir_def/types/unification.rs

```diff
@@ -0,0 +1,893 @@
+use std::borrow::Cow;
+
+use noirc_errors::Location;
+
+use crate::{
+    BinaryTypeOperator, Kind, Type, TypeBinding, TypeBindings, TypeVariable,
+    hir::type_check::TypeCheckError,
+    hir_def::{
+        expr::{HirCallExpression, HirExpression, HirIdent},
+        types,
+    },
+    node_interner::{ExprId, NodeInterner},
+};
+
+pub struct UnificationError;
+
+enum FunctionCoercionResult {
+    NoCoercion,
+    Coerced(Type),
+    UnconstrainedMismatch(Type),
+}
+
+/// When unifying types we sometimes need to adjust the algorithm a bit.
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum UnificationFlags {
+    /// Nothing special to do.
+    None,
+    /// If the right-hand side is `expr op constant`, don't try to move the constant to the left-hand side.
+    DoNotMoveConstantsOnTheRight,
+}
```

Comment:
Could maybe use this in the future to not call `Type::canonicalize` on every recursive invocation of try_unify as well

Reply:
Oh, you mean, when we call `try_unify` from, say `try_unify_by_moving_constant_terms`, there would be no need to call `canonicalize` on `lhs` and `rhs`, right?

Reply:
Even when we just call try_unify itself recursively on e.g. `a + (b + c) = d + (e + f)` we'll recur to `b + c = e + f` but we already canonicalized and don't need to canonicalize again every time